### DISCUSSION TITLE
Updating stimulus without end of current trial

### DISCUSSION THREAD

**Discussion by rMinion** [jspsych/jsPsych#1813] [Q&A]
Hi Guys, 

I have a conceptual problem. I want to update the stimulus without finishing the current trial. I have a set of trials that can be ended **only** by participant response (html-keyboard-response plugin).  

Periodically, in the random time interval (6-120s.) I want to change some particular element displayed on the screen without finishing the current trial. 

How can I achieve this? 

I tried `document.getElementById("myID").innerHTML = `<some value>``  within script tag but i get NULL. I think that it is connected to webpage elements loading order. 

I don't want to use `on_start` or `on_finish` within the trial mainly because it will not trigger the change of HTML when participants do not respond, and this is something that I want to avoid.

**Comment by cslongman** [comment]
I have done something similar to this (though the main difference is that I also get participants to respond twice).

In my experiment, I ended up using two timeline variables (one for the first response and one for the second response).

So I guess that in your experiment, you coudl do something similar (have more than one timeline variable) where you set a time limit on the first timeline object and if the second is needed you can display it and add the two RTs together to get a total RT.

I'm not sure if there is an easier way to do this unless you write your own plugin.

Hope this helps,

Cai

**Comment by becky-gilbert** [comment]
Hi @rMinion, the reason it doesn't work to put the line that changes the trial display inside of a script tag is because that code will run as soon as the page (experiment) first loads, whereas you need it to run during the appropriate trial.

In addition to @cslongman's great suggestion for setting this up as multiple trials to achieve the dipslay change, I thought I'd mention a few other options. One option is to add some code to the trial's `on_load` function that sets up a timer to change the display after some period of time. The trick with this solution is that you'll also want to cancel this timer when the trial ends, just in case the trial ends before the timer function runs. This is why we suggest using the [`jsPsych.pluginAPI.setTimeout` function](https://www.jspsych.org/core_library/jspsych-pluginAPI/#jspsychpluginapisettimeout), because this will register the timer with jsPsych, so that jsPsych can cancel all active timers at the end of a trial. The plugin that you're using might already cancel the timers at the end of the trial with the [`jsPsych.pluginAPI.clearAllTimeouts` function](https://www.jspsych.org/core_library/jspsych-pluginAPI/#jspsychpluginapiclearalltimeouts), but just in case it doesn't, you could add this in your trial's `on_finish` function. So it would look like this:
```js
// set up a function to get a random integer within a range of values
function randomNumberInRange(min, max) { 
    return Math.round(Math.random() * (max - min) + min);
} 

var trial = {
  // ... trial parameters ...
  on_load: function() {
    // get a random duration, in milliseconds, between 6 and 120 seconds
    var rand_duration = randomNumberInRange(6000, 120000);
    // set up a timer to change the display after this duration
    jsPsych.pluginAPI.setTimeout(function() {
        document.getElementById("myID").innerHTML = '';
    }, rand_duration);
  },
  on_finish: function(data) {
    jsPsych.pluginAPI.clearAllTimeouts();
  }
};
```
Another option is to modify the plugin, by adding the same lines of code as in the `on_load` and `on_finish` functions above. The `setTimeout` function should be called right away, when the trial loads, and the `clearAllTimeouts` function should be called when the trial ends (which usually happens inside of a function called `end_trial`). 

I hope this helps. Let us know if any of these solutions work for you, and if you need further guidance.
